cin_validator/rules/cin2022_23/rule_8770Q.py

```diff
@@ -0,0 +1,257 @@
+
+import pandas as pd
+
+from cin_validator.rule_engine import CINTable, RuleContext, RuleType, rule_definition
+from cin_validator.test_engine import run_rule
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+
+ChildIdentifiers = CINTable.ChildIdentifiers
+CINdetails = CINTable.CINdetails
+
+LAchildID = ChildIdentifiers.LAchildID
+PersonBirthDate = ChildIdentifiers.PersonBirthDate
+UPN = ChildIdentifiers.UPN
+UPNunknown = ChildIdentifiers.UPNunknown
+ReferralNFA = CINdetails.ReferralNFA
+
+# Reference date in header is needed to define the period of census.
+Header = CINTable.Header
+ReferenceDate = Header.ReferenceDate
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here, in place of 2885
+    code="8770Q",
+    rule_type=RuleType.QUERY,
+    # replace ChildProtectionPlans with the value in the module column of the excel sheet corresponding to this rule .
+    # Note that even if multiple tables are involved, one table will be named in the module column.
+    module=CINTable.ChildIdentifiers,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="Please check and either amend data or provide a reason: UPN or reason UPN missing expected for a child who is more than 5 years old",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[
+        UPNunknown,
+        UPNunknown,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace ChildProtectionPlans with the name of the table you need.
+    df_cid = data_container[ChildIdentifiers].copy()
+    df_cin = data_container[CINdetails].copy()
+
+    # Before you begin, rename the index so that the initial row positions can be kept intact.
+    df_cid.index.name = "ROW_ID"
+    df_cin.index.name = "ROW_ID"
+
+    # Resetting the index causes the ROW_IDs to become columns of their respective DataFrames
+    # so that they can come along when the merge is done.
+    df_cid.reset_index(inplace=True)
+    df_cin.reset_index(inplace=True)
+
+    # get collection period
+    header = data_container[Header]
+    ref_date = header[ReferenceDate].iloc[0]
+    ref_date_minus6 = ref_date - pd.DateOffset(years=6)
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # If <ReferralNFA> (N00112) = 0 or false
+    # and <PersonBirthDate> (N00066) is earlier than [<ReferenceDate> (N00603) minus 6 years]
+    # then either <UPN> (N00001) or a valid <UPNunknown> (N00135) should be present
+
+    # get only CINdetails where ReferralNFA is 0 or false
+    df_cin = df_cin[
+        (df_cin[ReferralNFA] == "0") | (df_cin[ReferralNFA].str.lower() == "false")
```

Comment:
It's good to sanitise the input of false here! Another rule picks up if the value isn't 0 or false as expected, but this doersn't hurt, nice!